### PR TITLE
Add beanhub-cli and deps

### PR DESCRIPTION
**Pull request by fangpenlin**
## Description of changes

[beanhub-cli](https://github.com/LaunchPlatform/beanhub-cli) brings many useful beancount command tools powered by:
- [beanhub-forms](https://github.com/LaunchPlatform/beanhub-forms)
- [beanhub-import](https://github.com/LaunchPlatform/beanhub-import)
- [beanhub-extract](https://github.com/LaunchPlatform/beanhub-extract)

depends on: #310502

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by kirillrdy** [comment]
if its still WIP, can you Convert to Draft so that people don't get notified

**Comment by fangpenlin** [comment]
@kirillrdy hey, sorry. next time I will open it as DRAFT if it's still working in progress. It was waiting #310502 to merged as this PR depends on it. Now the PR was merged, I just rebased and pushed. This PR is ready for review.

File: pkgs/development/python-modules/beanhub-cli/default.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+, click
+, uvicorn
+, starlette-wtf
+, beancount-parser
+, beancount-black
+, beanhub-import
+, beanhub-forms
+, poetry-core
+, httpx
+, pytestCheckHook
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-cli/default.nix:22]: I'd order the python modules ASCII-betically, and keep `pytestCheckHook` with the other packaging stuff.

File: pkgs/development/python-modules/beanhub-cli/default.nix [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  src = fetchFromGitHub {
+    owner = "LaunchPlatform";
+    repo = "beanhub-cli";
+    rev = "refs/tags/${version}";
+    hash = "sha256-vYBbaUVJAs+aIp6aQpJb62DEDxe/sQTzgOkjPq6ADoc=";
+  };
+
+  buildInputs = [
+    poetry-core
+  ];
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-cli/default.nix:41]: The input names seem to have changed, if I read [the docs](https://nixos.org/manual/nixpkgs/unstable/#buildpythonpackage-function) correctly this should be `build-system`.

File: pkgs/development/python-modules/beanhub-cli/default.nix [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    rich
+    starlette-wtf
+    fastapi
+    click
+    uvicorn
+    beancount-parser
+    beancount-black
+    beanhub-import
+    beanhub-forms
+  ];
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-cli/default.nix:59]: Similarly this should be `dependencies`.

Also please sort the list.

File: pkgs/development/python-modules/beanhub-cli/default.nix [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  pythonImportsCheck = [
+    "beanhub_cli"
+  ];
+
+  meta = with lib; {
+    description = "Command line tools for BeanHub or Beancount users";
+    homepage = "https://github.com/LaunchPlatform/beanhub-cli/";
+    changelog = "https://github.com/LaunchPlatform/beanhub-cli/releases/tag/${version}";
+    license = with licenses; [ mit ];
+    maintainers = with maintainers; [ ambroisie ];
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-cli/default.nix:75]: I don't want to maintain this package, I don't make use of it.

Please add yourself to the maintainer list (in a commit named `maintains: add <handle>` beforein accordance with the [contributing docs](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) :-).

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -1461,6 +1461,14 @@ self: super: with self; {
...
   beancount-parser = callPackage ../development/python-modules/beancount-parser { };
 
+  beanhub-extract = callPackage ../development/python-modules/beanhub-extract { };
+
+  beanhub-import = callPackage ../development/python-modules/beanhub-import { };
+
+  beanhub-forms = callPackage ../development/python-modules/beanhub-forms { };
+
+  beanhub-cli = callPackage ../development/python-modules/beanhub-cli { };
+
```
Comment (ambroisie) [inline pkgs/top-level/python-packages.nix:1471]: Please sort these.

File: pkgs/development/python-modules/beanhub-extract/default.nix [hunk 6]
```diff
@@ -0,0 +1,47 @@
+{ lib
+, fetchFromGitHub
+, buildPythonPackage
+, pythonOlder
+, pytz
+, poetry-core
+, pytestCheckHook
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-extract/default.nix:7]: Sorting

File: pkgs/development/python-modules/beanhub-extract/default.nix [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  src = fetchFromGitHub {
+    owner = "LaunchPlatform";
+    repo = "beanhub-extract";
+    rev = "refs/tags/${version}";
+    hash = "sha256-Wt8ZCyCaERNXEd0/Q89QWUW/wGFSHAP2RZLhnv5xkgY=";
+  };
+
+  buildInputs = [
+    poetry-core
+  ];
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-extract/default.nix:26]:
```suggestion
  build-system = [
    poetry-core
  ];
```

File: pkgs/development/python-modules/beanhub-forms/default.nix [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  src = fetchFromGitHub {
+    owner = "LaunchPlatform";
+    repo = "beanhub-forms";
+    rev = "refs/tags/${version}";
+    hash = "sha256-313c+ENmTe1LyfEiMXNB9AUoGx3Yv/1D0T3HnAbd+Zw=";
+  };
+
+  buildInputs = [
+    poetry-core
+  ];
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-forms/default.nix:30]:
```suggestion
  build-system = [
    poetry-core
  ];
```

File: pkgs/development/python-modules/beanhub-import/default.nix [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  src = fetchFromGitHub {
+    owner = "LaunchPlatform";
+    repo = "beanhub-import";
+    rev = "refs/tags/${version}";
+    hash = "sha256-BH+d7cKC83b7T2Z1Xlwpsv3KACo1vDn6v65tElhir/c=";
+  };
+
+  buildInputs = [
+    poetry-core
+  ];
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-import/default.nix:32]:
```suggestion
  build-system = [
    poetry-core
  ];
```

File: pkgs/development/python-modules/beanhub-cli/default.nix (hunk 2 above)
Comment (ambroisie) [reply pkgs/development/python-modules/beanhub-cli/default.nix:41]:
```suggestion
  build-system = [
    poetry-core
  ];
```

File: pkgs/development/python-modules/beanhub-cli/default.nix (hunk 3 above)
Comment (ambroisie) [reply pkgs/development/python-modules/beanhub-cli/default.nix:59]:
```suggestion
  dependencies = [
    beancount-black
    beancount-parser
    beanhub-forms
    beanhub-import
    click
    fastapi
    jinja2
    pydantic
    pydantic-core
    pydantic-settings
    pytz
    pyyaml
    rich
    starlette-wtf
    uvicorn
  ];
```

File: pkgs/development/python-modules/beanhub-extract/default.nix [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+  propagatedBuildInputs = [
+    pytz
+  ];
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-extract/default.nix:30]:
```suggestion
  dependencies = [
    pytz
  ];
```

File: pkgs/development/python-modules/beanhub-forms/default.nix [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+  propagatedBuildInputs = [
+    pydantic
+    pyyaml
+    wtforms
+    jinja2
+  ];
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-forms/default.nix:37]:
```suggestion
  dependencies = [
    jinja2
    pydantic
    pyyaml
    wtforms
  ];
```

File: pkgs/development/python-modules/beanhub-forms/default.nix [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+, fetchFromGitHub
+, buildPythonPackage
+, pythonOlder
+, pydantic
+, pyyaml
+, wtforms
+, jinja2
+, multidict
+, poetry-core
+, pytestCheckHook
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-forms/default.nix:11]:
```suggestion
, pytestCheckHook
, jinja2
, multidict
, poetry-core
, pydantic
, pyyaml
, wtforms
```

File: pkgs/development/python-modules/beanhub-import/default.nix [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+, pythonOlder
+, pytz
+, pydantic
+, pyyaml
+, jinja2
+, beancount-parser
+, beancount-black
+, beanhub-extract
+, poetry-core
+, pytestCheckHook
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-import/default.nix:13]:
```suggestion
, pytestCheckHook
, beancount-black
, beancount-parser
, beanhub-extract
, jinja2
, poetry-core
, pydantic
, pytz
, pyyaml
```

File: pkgs/development/python-modules/beanhub-import/default.nix [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+  propagatedBuildInputs = [
+    pytz
+    pydantic
+    pyyaml
+    jinja2
+    beancount-parser
+    beancount-black
+    beanhub-extract
+  ];
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-import/default.nix:42]:
```suggestion
  dependencies = [
    beancount-black
    beancount-parser
    beanhub-extract
    jinja2
    pydantic
    pytz
    pyyaml
  ];
```

File: pkgs/development/python-modules/beanhub-import/default.nix [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+  nativeCheckInputs = [
+    pytestCheckHook
+  ];
+
+  pythonImportsCheck = [
+    "beanhub_import"
+  ];
+
+  meta = with lib; {
+    description = "Beanhub-import is a simple, declarative, smart, and easy-to-use library for importing extracted transactions from beanhub-extract. It generates Beancount transactions based on predefined rules.";
```
Comment (ambroisie) [inline pkgs/development/python-modules/beanhub-import/default.nix:53]: Nit: this description is a bit long IMO.

File: pkgs/development/python-modules/beanhub-cli/default.nix (hunk 1 above)
Comment (ambroisie) [reply pkgs/development/python-modules/beanhub-cli/default.nix:22]:
```suggestion
, pytestCheckHook
, beancount-black
, beancount-parser
, beanhub-forms
, beanhub-import
, click
, fastapi
, httpx
, jinja2
, poetry-core
, pydantic
, pydantic-core
, pydantic-settings
, pytz
, pyyaml
, rich
, starlette-wtf
, uvicorn
```

**Review by ambroisie — CHANGES_REQUESTED** [review]
Thanks for adding the packages, a few nits mostly regarding evolutions `buildPythonPackage`'s API and general formatting.

File: pkgs/development/python-modules/beanhub-cli/default.nix (hunk 1 above)
Comment (fangpenlin) [reply pkgs/development/python-modules/beanhub-cli/default.nix:22]: @ambroisie hey, thanks! sorry I am pretty new to contributing to nixpkgs. I usually run

```
nix fmt .
```

for my own nix projects. I just tried to run it but it says

```
error: flake 'git+file:///home/fangpen/workspace/playground/nixpkgs' does not provide attribute 'formatter.x86_64-linux'
```

ideally I think there should be a tool for this. but if not it's okay, thanks for calling it out anyway 🙇‍♀️

File: pkgs/development/python-modules/beanhub-cli/default.nix (hunk 4 above)
Comment (fangpenlin) [reply pkgs/development/python-modules/beanhub-cli/default.nix]: ah, sorry, didn't meant to throw these packages at your, forgot to add myself as an author 😅

**Comment by fangpenlin** [comment]
@ambroisie hey, sorry for pulling you into this unrelative PR, thanks again for the detailed review. I believe I have addressed all the feedbacks. Please let me know if there's anything else I can improve 🙏

**Review by nicoonoclaste — APPROVED** [review]

File: pkgs/development/python-modules/beanhub-cli/default.nix [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+, pyyaml
+, rich
+, starlette-wtf
+, uvicorn
+}:
+
+buildPythonPackage rec {
+  pname = "beanhub-cli";
+  version = "1.1.3";
+  format = "pyproject";
```
Comment (kirillrdy) [inline pkgs/development/python-modules/beanhub-cli/default.nix:28]:
```suggestion
  pyproject = true'
```

File: pkgs/development/python-modules/beanhub-extract/default.nix [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+, pythonOlder
+, pytestCheckHook
+, poetry-core
+, pytz
+}:
+
+buildPythonPackage rec {
+  pname = "beanhub-extract";
+  version = "0.0.7";
+  format = "pyproject";
```
Comment (kirillrdy) [inline pkgs/development/python-modules/beanhub-extract/default.nix:13]:
```suggestion
  pyproject = true;
```

File: pkgs/development/python-modules/beanhub-forms/default.nix [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+, poetry-core
+, pydantic
+, pyyaml
+, wtforms
+}:
+
+buildPythonPackage rec {
+  pname = "beanhub-forms";
+  version = "0.1.3";
+  format = "pyproject";
```
Comment (kirillrdy) [inline pkgs/development/python-modules/beanhub-forms/default.nix:17]:
```suggestion
  pyproject = true;
```

File: pkgs/development/python-modules/beanhub-import/default.nix [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+, poetry-core
+, pydantic
+, pytz
+, pyyaml
+}:
+
+buildPythonPackage rec {
+  pname = "beanhub-import";
+  version = "0.1.5";
+  format = "pyproject";
```
Comment (kirillrdy) [inline pkgs/development/python-modules/beanhub-import/default.nix:19]:
```suggestion
  pyproject = true;
```

File: pkgs/development/python-modules/beanhub-cli/default.nix [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+  nativeCheckInputs = [
+    pytestCheckHook
+    httpx
+  ];
+
+  pythonImportsCheck = [
+    "beanhub_cli"
+  ];
+
+  meta = with lib; {
```
Comment (kirillrdy) [inline pkgs/development/python-modules/beanhub-cli/default.nix:70]: please consider adding https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

File: pkgs/development/python-modules/beanhub-extract/default.nix [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+
+  nativeCheckInputs = [
+    pytestCheckHook
+  ];
+
+  pythonImportsCheck = [
+    "beanhub_extract"
+  ];
+
+  meta = with lib; {
```
Comment (kirillrdy) [inline pkgs/development/python-modules/beanhub-extract/default.nix:40]: https://github.com/NixOS/nixpkgs/issues/208242

**Review by kirillrdy — COMMENTED** [review]
Result of `nixpkgs-review pr 310655` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python312Packages.beanhub-cli</li>
    <li>python312Packages.beanhub-cli.dist</li>
    <li>python312Packages.beanhub-import</li>
    <li>python312Packages.beanhub-import.dist</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>beanhub-cli (python311Packages.beanhub-cli)</li>
    <li>beanhub-cli.dist (python311Packages.beanhub-cli.dist)</li>
    <li>python311Packages.beanhub-extract</li>
    <li>python311Packages.beanhub-extract.dist</li>
    <li>python311Packages.beanhub-forms</li>
    <li>python311Packages.beanhub-forms.dist</li>
    <li>python311Packages.beanhub-import</li>
    <li>python311Packages.beanhub-import.dist</li>
    <li>python312Packages.beanhub-extract</li>
    <li>python312Packages.beanhub-extract.dist</li>
    <li>python312Packages.beanhub-forms</li>
    <li>python312Packages.beanhub-forms.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/beanhub-extract/default.nix (hunk 21 above)
Comment (ambroisie) [reply pkgs/development/python-modules/beanhub-extract/default.nix:40]: I agree with Sandro here, it's fine if it's limited to a small scope (here `meta`, but I also find it okay for `options`). 👎 on changing it IMO.

Comment (kirillrdy) [reply pkgs/development/python-modules/beanhub-extract/default.nix:40]: I don't have strong opinion here. I merely presenting people with information

**Comment by fangpenlin** [comment]
thanks for the review feedbacks! I will address them later today

**Comment by fangpenlin** [comment]
@kirillrdy all the feedbacks addressed, please have a look again and let me know if there's anything else I can improve 🙏

**Review by kirillrdy — APPROVED** [review]
Result of `nixpkgs-review pr 310655` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>beanhub-cli (python311Packages.beanhub-cli)</li>
    <li>beanhub-cli.dist (python311Packages.beanhub-cli.dist)</li>
    <li>python311Packages.beanhub-extract</li>
    <li>python311Packages.beanhub-extract.dist</li>
    <li>python311Packages.beanhub-forms</li>
    <li>python311Packages.beanhub-forms.dist</li>
    <li>python311Packages.beanhub-import</li>
    <li>python311Packages.beanhub-import.dist</li>
    <li>python312Packages.beanhub-cli</li>
    <li>python312Packages.beanhub-cli.dist</li>
    <li>python312Packages.beanhub-extract</li>
    <li>python312Packages.beanhub-extract.dist</li>
    <li>python312Packages.beanhub-forms</li>
    <li>python312Packages.beanhub-forms.dist</li>
    <li>python312Packages.beanhub-import</li>
    <li>python312Packages.beanhub-import.dist</li>
  </ul>
</details>

**Comment by fangpenlin** [comment]
@ambroisie can you please review again when you have time? 🙏

**Review by ambroisie — APPROVED** [review]

**Pull request merged**